RedHatInsights/rhsm-subscriptions
# SWATCH-969: Replace ProductId enum with class that checks against product-config

Jira issue: [SWATCH-969](https://issues.redhat.com/browse/SWATCH-969)

Description
===========

Remove productId enum in favor of a class that validates against product-config (extremely similar
to #2464). I made these changes on top of that PR in order to avoid merge conflicts, so #2464 must
land first.

Testing
=======

Setup
-----

1. Deploy the service:

```shell
DEV_MODE=true \
  ./gradlew :bootRun
```

2. Ensure opted in:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/opt-in \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Verification - Happy Path
-------------------------

Try the following APIs, expecting successful responses...

capacity (deprecated):

```shell
http ':8000/api/rhsm-subscriptions/v1/capacity/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

capacity:

```shell
http ':8000/api/rhsm-subscriptions/v1/capacity/products/RHEL for x86/Sockets' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

hosts:

```shell
http ':8000/api/rhsm-subscriptions/v1/hosts/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

instances:

```shell
http ':8000/api/rhsm-subscriptions/v1/instances/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

subscriptions:

```shell
http ':8000/api/rhsm-subscriptions/v1/subscriptions/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally (deprecated):

```shell
http ':8000/api/rhsm-subscriptions/v1/tally/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally:

```shell
http ':8000/api/rhsm-subscriptions/v1/tally/products/RHEL for x86/Sockets' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Verification - Bad Requests
---------------------------

Try the following APIs, expected 404 w/ an error message explaining the missing productId...

capacity (deprecated):

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

capacity:

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/NotAProduct/Cores \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

hosts:

```shell
http :8000/api/rhsm-subscriptions/v1/hosts/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

instances:

```shell
http :8000/api/rhsm-subscriptions/v1/instances/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

subscriptions:

```shell
http :8000/api/rhsm-subscriptions/v1/subscriptions/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally (deprecated):

```shell
http :8000/api/rhsm-subscriptions/v1/tally/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally:

```shell
http :8000/api/rhsm-subscriptions/v1/tally/products/NotAProduct/Cores \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Draft until #2464 lands.
